AcademyApp/ViewControllers/LoginViewController.swift

```diff
@@ -0,0 +1,226 @@
+    @IBOutlet weak var doneButton: RoundedCornersButton!
+    @IBOutlet weak var showButton: UIButton!
+    @IBOutlet weak var scrollView: UIScrollView!
+    
+    // MARK: - @IBActions
+    
+    @IBAction func backgroundTapped(_ sender: UITapGestureRecognizer) {
+        self.view.endEditing(true)
+    }
+    
+    @IBAction func emailTextFieldEditingDidChange(_ sender: UITextField) {
+        
+        guard let emailText = emailTextField.text, let passwordText = passwordTextField.text  else {
+            return
+        }
+        
+        doneButton.isEnabled = String.isValidEmail(possibleEmail: emailText) && passwordText.count > 0 ? true : false
+    }
+    
+    @IBAction func passwordTextFieldEditingDidChange(_ sender: UITextField) {
+        
+        guard let emailText = emailTextField.text, let passwordText = passwordTextField.text  else {
+            return
+        }
+        
+        
+        UIView.animate(withDuration: 0.2) {
+            self.showButton.isHidden = passwordText.count > 0 ? false : true
+        }
+        
+        doneButton.isEnabled = String.isValidEmail(possibleEmail: emailText) && passwordText.count > 0 ? true : false
+    }
+    
+    @IBAction func showButtonTapped(_ sender: UIButton) {
+        passwordTextField.isSecureTextEntry = !passwordTextField.isSecureTextEntry
+        setTitleForShowButton(for: passwordTextField.isSecureTextEntry)
+    }
+    
+}
+
+// MARK: - UITextField's delegate methods
+
+extension LoginViewController : UITextFieldDelegate
+{
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        self.view.endEditing(true)
+        return true
+    }
+}
+
+// MARK: - UI methods
+
+private extension LoginViewController
+{
+    
+    func setTitleForShowButton(for secureTextState: Bool){
+        
+        let attributes = showButton.attributedTitle(for: .normal)?.attributes(at: 0, effectiveRange: nil)
+        let title = secureTextState ? "SHOW" : "HIDE"
+        
+        showButton.setAttributedTitle(NSAttributedString(string: title, attributes: attributes), for: .normal)
+    }
+}
+
+// MARK: - Responding to keyboard notifications
+
+private extension LoginViewController
+{
+   
+    @objc func keyboardWillShow(notification: Notification){
+        
+        setKeyboardIsVisible(true, with: notification)
+    }
+    
+    @objc func keyboardWillHide(notification: Notification){
+        
+        setKeyboardIsVisible(false, with: notification)
+    }
+    
+    func setKeyboardIsVisible(_ isVisible : Bool, with notification: Notification){
+        
+        guard let keyboardRect = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue  else {
+            return
+        }
+        
+        scrollViewBottomConstraint.constant = isVisible ?  -keyboardRect.height  : 0
+        doneButtonLoginFormVerticalSpacingConstraint.constant = isVisible ? 56 : defaultDoneButtonLoginFormVerticalSpacingConstant
```

Comment:
avoid magic numbers, you can create a private enum with constants

class ViewController: UIViewController {
    private enum Config {
        static let someSpacing: CGFloat = 56
    }
}

Reply:
Moved number literals to struct Config.